### PR TITLE
[SPARK-30315][SQL] Add adaptive execution context

### PR DESCRIPTION
**Pull request by maryannxue**
### What changes were proposed in this pull request?
This is a minor code refactoring PR. It creates an adaptive execution context class to wrap objects shared across main query and sub-queries.

### Why are the changes needed?
This refactoring will improve code readability and reduce the number of parameters used to initialize `AdaptiveSparkPlanExec`.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Passed existing UTs.

### CODE REVIEW

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 1]
```diff
@@ -59,10 +59,8 @@ import org.apache.spark.util.ThreadUtils
  */
 case class AdaptiveSparkPlanExec(
     initialPlan: SparkPlan,
-    @transient session: SparkSession,
+    @transient adaptiveExecutionContext: AdaptiveExecutionContext,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala]: for simplicity, we can just name it `context`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 2]
```diff
@@ -275,7 +275,7 @@ object QueryExecution {
     Seq(
       // `AdaptiveSparkPlanExec` is a leaf node. If inserted, all the following rules will be no-op
       // as the original plan is hidden behind `AdaptiveSparkPlanExec`.
-      InsertAdaptiveSparkPlan(sparkSession),
+      InsertAdaptiveSparkPlan(AdaptiveExecutionContext(sparkSession)),
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: It seems clearer to me to create `AdaptiveExecutionContext` inside `InsertAdaptiveSparkPlan` rule, and just pass `SparkSession` to the rule.

**Comment by cloud-fan** [comment]
retest it please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Review by cloud-fan — APPROVED** [review]

**Review by gatorsmile — APPROVED** [review]
Thanks! Merged to master.

**Pull request closed without merging**